quinn-proto/src/packet.rs

```diff
@@ -52,19 +54,51 @@ impl PartialDecode {
         }
     }
 
+    pub fn is_handshake(&self) -> bool {
+        match self.invariant_header {
+            InvariantHeader::Long {
+                version: VERSION,
+                first,
+                ..
+            } => match PacketType::from_byte(first).unwrap() {
+                PacketType::Initial => true,
+                PacketType::Long(LongType::Handshake) => true,
+                _ => false,
+            },
+            InvariantHeader::Long { .. } => false,
+            InvariantHeader::Short { .. } => false,
+        }
+    }
+
     pub fn dst_cid(&self) -> ConnectionId {
         self.invariant_header.dst_cid()
     }
 
-    pub fn finish(self) -> Result<(Packet, Option<BytesMut>), PacketDecodeError> {
+    pub fn finish(
+        self,
+        pn_key: &PacketNumberKey,
+    ) -> Result<(Packet, Option<BytesMut>), PacketDecodeError> {
         let Self {
             invariant_header,
             mut buf,
         } = self;
         let (payload_len, header, allow_coalesced) = match invariant_header {
             InvariantHeader::Short { first, dst_cid } => {
                 let key_phase = first & KEY_PHASE_BIT != 0;
-                let number = PacketNumber::decode(&mut buf)?;
+                if buf.remaining() < 1 {
+                    return Err(PacketDecodeError::InvalidHeader(
+                        "cannot read packet number",
```

Comment:
"can't read" suggests an unspecified internal error to me; perhaps rephrase as "missing"?

Reply:
I've changed it to `header ends before packet number`.